src/main/java/com/scalar/db/storage/jdbc/JDBC.java

```diff
@@ -0,0 +1,187 @@
+import com.scalar.db.storage.jdbc.query.QueryBuilder;
+import org.apache.commons.dbcp2.BasicDataSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * A storage implementation with JDBC for {@link DistributedStorage}.
+ *
+ * <p>Note that the consistency in an operation is always LINEARIZABLE in this implementation. Even
+ * if consistency is specified in an operation, it will be ignored.
+ *
+ * @author Toshihiro Suzuki
+ */
+@ThreadSafe
+public class JDBC implements DistributedStorage {
+  private static final Logger LOGGER = LoggerFactory.getLogger(JDBC.class);
+
+  private final BasicDataSource dataSource;
+  private final JDBCService jdbcService;
+  @Nullable private String defaultSchema;
+  @Nullable private String defaultTable;
+
+  @Inject
+  public JDBC(DatabaseConfig config) {
+    dataSource = JDBCUtils.initDataSource(config);
+
+    String schemaPrefix = config.getNamespacePrefix().orElse("");
+
+    TableMetadataManager tableMetadataManager = new TableMetadataManager(dataSource, schemaPrefix);
+    OperationChecker operationChecker = new OperationChecker(tableMetadataManager);
+    QueryBuilder queryBuilder =
+        new QueryBuilder(
+            tableMetadataManager, JDBCUtils.getRDBType(config.getContactPoints().get(0)));
+    jdbcService = new JDBCService(operationChecker, queryBuilder, schemaPrefix);
+  }
+
+  @VisibleForTesting
+  JDBC(BasicDataSource dataSource, JDBCService jdbcService) {
+    this.dataSource = dataSource;
+    this.jdbcService = jdbcService;
+  }
+
+  @Override
+  public void with(String namespace, String tableName) {
+    defaultSchema = namespace;
+    defaultTable = tableName;
+  }
+
+  @Override
+  public void withNamespace(String namespace) {
+    defaultSchema = namespace;
+  }
+
+  @Override
+  public Optional<String> getNamespace() {
+    return Optional.ofNullable(defaultSchema);
+  }
+
+  @Override
+  public void withTable(String tableName) {
+    defaultTable = tableName;
+  }
+
+  @Override
+  public Optional<String> getTable() {
+    return Optional.ofNullable(defaultTable);
+  }
+
+  @Override
+  public Optional<Result> get(Get get) throws ExecutionException {
+    try (Connection connection = dataSource.getConnection()) {
+      return jdbcService.get(get, connection, defaultSchema, defaultTable);
+    } catch (SQLException e) {
+      throw new ExecutionException("An error occurred", e);
+    }
+  }
+
+  @Override
+  public Scanner scan(Scan scan) throws ExecutionException {
+    Connection connection = null;
+    try {
+      connection = dataSource.getConnection();
+      return jdbcService.scan(scan, connection, defaultSchema, defaultTable);
+    } catch (SQLException e) {
+      try {
+        if (connection != null) {
```

Comment:
If there is a chance of `null` for connection`,  I think it should be checked when created. 
https://github.com/scalar-labs/scalardb/pull/133/files#diff-269c4117791939509f752c12912fd795785d698c30c82d92b05a54bbc29c62f5R105

Reply:
Sorry, I misunderstood the code. Based on the chat discussion, it's not needed to be handled.